server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeDocumentsSorter.java

```diff
@@ -45,19 +49,27 @@ protected void setPivot(int i) {
             protected int comparePivot(int j) {
                 Long[] currentDimensions = dimensionsReader.apply(j);
                 for (int i = dimensionId + 1; i < dimensions.length; i++) {
-                    Long dimension = currentDimensions[i];
-                    if (!Objects.equals(dimensions[i], dimension)) {
-                        if (dimensions[i] == null && dimension == null) {
+                    Dimension dimension = dimensionsOrder.get(i);
+                    Long dimensionValue = currentDimensions[i];
+                    if (!Objects.equals(dimensions[i], dimensionValue)) {
+                        if (dimensions[i] == null && dimensionValue == null) {
                             return 0;
                         }
-                        if (dimension == null) {
+                        if (dimensionValue == null) {
                             return -1;
                         }
                         if (dimensions[i] == null) {
                             return 1;
                         }
-                        return Long.compare(dimensions[i], dimension);
+                        if (dimension instanceof NumericDimension) {
+                            NumericDimension numericDimension = (NumericDimension) dimension;
+                            if (numericDimension.isUnsignedLong()) {
```

Comment:
Lets avoid if else checks, lets have a default implementation of comparator in the dimension base class, and child dimensions [such as unsignedLongDimension] can overload if needed.

Reply:
We now have an ENUM - `ComparatorType`, which can have 2 values `LONG`and `UNSIGNED_LONG`and they explicitly perform their respective comparisons.
The `Dimension` interface has a default method - `getComparatorType` which returns the `ComparatorType.LONG` enum value. Subclasses can override this method to return the appropriate enum value as needed.